### DISCUSSION TITLE
bug: `TransportConfig` not assignable to ethers `ExternalProvider | JsonRpcFetchFunc`

### DISCUSSION THREAD

**Discussion by dtbuchholz** [wevm/viem#792] [Question] [answered]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package Version

1.1.7

### Current Behavior

Following the [adapters guide](https://wagmi.sh/core/ethers-adapters#wallet-client--signer) (with slight edits) for `viem`<>`ethers` conversion to a `Signer`, I receive the following error about the `transport` variable that gets destructured from a `WalletClient`:
```md
Argument of type 'TransportConfig<string, EIP1193RequestFn> & Record<string, any>' is not assignable to parameter of type 'ExternalProvider | JsonRpcFetchFunc'.
  Type 'TransportConfig<string, EIP1193RequestFn> & Record<string, any>' is not assignable to type 'ExternalProvider'.
    Types of property 'request' are incompatible.
      Type 'EIP1193RequestFn' is not assignable to type '(request: { method: string; params?: any[]; }) => Promise<any>'.
        Types of parameters 'args' and 'request' are incompatible.
          Type '{ method: string; params?: any[]; }' is not assignable to type '{ method: unknown; params: unknown; }'.
            Property 'params' is optional in type '{ method: string; params?: any[]; }' but required in type '{ method: unknown; params: unknown; }'.ts(2345)
```
The exact code is the following, which is a hook for a React app:
```ts
import * as React from "react";
import { type WalletClient, useWalletClient } from "wagmi";
import { providers, type Signer } from "ethers";

function walletClientToSigner(walletClient: WalletClient): Signer {
  const { account, chain, transport } = walletClient;
  const network = {
    chainId: chain.id,
    name: chain.name,
    ensAddress: chain.contracts?.ensRegistry?.address,
  };
  const provider = new providers.Web3Provider(transport, network); // Issue here
  const signer = provider.getSigner(account.address);
  return signer;
}

export function useSigner({ chainId }: { chainId?: number } = {}):
  | Signer
  | undefined {
  const { data: walletClient } = useWalletClient({ chainId });
  return React.useMemo(
    () => (walletClient ? walletClientToSigner(walletClient) : undefined),
    [walletClient]
  );
}
```
In order to resolve this issue, the following is adjusted in `walletClientToSigner` but no longer makes use of `transport`:
```ts
const { account, chain } = walletClient;
const network = {
    chainId: chain.id,
    name: chain.name,
    ensAddress: chain.contracts?.ensRegistry?.address,
};
const rpcUrl = chain.rpcUrls.default.http[0];
const provider = new providers.JsonRpcProvider(rpcUrl, network);
```

### Expected Behavior

Ability to drop in `ethers` adapters example and use `wagmi`/`rainbowkit` without issue (needed for libraries that use `ethers` as a dep).

### Steps To Reproduce

1. Create a React app and install the following: `ethers@^5.7.2`, `viem@^1.1.7`, `wagmi@~1.2.0` (version needed for rainbowkit), `@rainbow-me/rainbowkit@^1.0.3`
2. Create a `useSigner.ts` hook that uses the code noted above.

You'll get the error about `transport`. Note that I've tried the adapter guide for both `ethers` v5 and v6; I'm using v5 here for lib dep purposes but the `transport` issue is thrown in both. I've also tried various version of both `wagmi` (~1.2.0 to latest) and `viem` (~1.0.6 to latest)without success.
![Screenshot 2023-06-26 at 4 24 19 PM](https://github.com/wagmi-dev/viem/assets/13358940/1ec54e27-b976-43f7-bfe7-a5217a861771)

### Anything else?

Also note that in your docs, _I think_ the `ethers` v5 could make use of `wagmi` and doesn't need `@wagmi/core`. See the original snippet, where I use the following:
```ts
import { type WalletClient, useWalletClient } from "wagmi";
// The import above is used and works, instead of needing the following
import { type WalletClient, getWalletClient } from '@wagmi/core'
```
Lastly, the docs ethers v6 docs are incorrect and use `getWalletClient` instead of `useWalletClient` in `getEthersSigner`.
![Screenshot 2023-06-26 at 4 22 38 PM](https://github.com/wagmi-dev/viem/assets/13358940/41e23d2a-7b33-46bf-85e2-109d561c7ec8)

**Comment by prakhar728** [comment]
One of the workaround I found was using `const rpcUrl = chain.rpcUrls.default.http[0];` and then passing rpcUrl here 
` const provider = new providers.JsonRpcProvider(rpcUrl, network);`.
Definitely would be cool if we could use transport directly?

**Comment by jxom** [comment]
The ethers type def is wrong, you can just ts-ignore it and it should still work.

**Reply by dtbuchholz** [reply, marked as answer]
@jxom Ugh yeah, that fixed it...so simple! Alternatively, I came across something in the docs that does spell out a config that resolves the issue, too, in case it ever comes up again: [here](https://wagmi.sh/react/typescript). Namely, `tsconfig.json` needs the following:
```json
{
  "compilerOptions": {
    "strict": true
  }
}
```
Once this is set, the issue with the transport type error no longer occurs.

**Reply by MeysamAzari** [reply]
i have the same problem in react javascript. what should i do?

**Reply by dtbuchholz** [reply]
@MeysamAzari I have an example React JS template of what I implemented [here](https://github.com/tablelandnetwork/react-js-tableland-template). Take a look at the custom `useSigner` hook, which converts viem->ethers. There's some other setup I've done with `wagmi.js` and `providers.jsx`, for context, and the example component uses the hook. Also, if you're using Next and/or TS, there are also similar template repos within that org.
